src/async_traits.rs

```diff
@@ -48,40 +62,183 @@ where
+
+    /// Issues a function `f` as a transaction.
+    ///
+    /// If it fails, asynchronously calls `retry` to decide if to retry.
+    #[cfg(feature = "cockroach")]
+    async fn transaction_async_with_retry<R, Func, Fut, RetryFut, RetryFunc, 'a>(
+        &'a self,
+        f: Func,
+        retry: RetryFunc,
+    ) -> Result<R, DieselError>
+    where
+        R: Send + 'static,
+        Fut: Future<Output = Result<R, DieselError>> + Send,
+        Func: Fn(SingleConnection<Conn>) -> Fut + Send + Sync,
+        RetryFut: Future<Output = bool> + Send,
+        RetryFunc: Fn() -> RetryFut + Send + Sync,
+    {
+        // Check out a connection once, and use it for the duration of the
+        // operation.
+        let conn = Arc::new(self.get_owned_connection());
+
+        // Refer to CockroachDB's guide on advanced client-side transaction
+        // retries for the full context here. In short, they expect a particular
+        // name for this savepoint, but Diesel has Opinions on savepoint names,
+        // so we use this session variable to identify that any name is valid.
+        //
+        // TODO: It may be preferable to set this once per connection -- but
+        // that'll require more interaction with how sessions with the database
+        // are constructed.
+        Self::start_transaction(&conn).await?;
+        conn.run_with_shared_connection(|conn| {
+            conn.batch_execute("SET LOCAL force_savepoint_restart = true")
+        })
+        .await?;
+
+        loop {
+            // Add a SAVEPOINT to which we can later return.
+            Self::add_retry_savepoint(&conn).await?;
+
+            let async_conn = SingleConnection(Self::as_async_conn(&conn).0.clone());
+            match f(async_conn).await {
+                Ok(value) => {
+                    // The user-level operation succeeded: try to commit the
+                    // transaction by RELEASE-ing the retry savepoint.
+                    if let Err(err) = Self::commit_transaction(&conn).await {
+                        // We're still in the transaction, but we at least
+                        // tried to ROLLBACK to our savepoint.
+                        if !retryable_error(&err) || !retry().await {
+                            // Bail: ROLLBACK the initial BEGIN statement too.
+                            let _ = Self::rollback_transaction(&conn).await;
+                            return Err(err);
+                        }
+                        // ROLLBACK happened, we want to retry.
+                        continue;
+                    }
+
+                    Self::commit_transaction(&conn).await?;
+                    return Ok(value);
+                }
+                Err(user_error) => {
+                    // The user-level operation failed: ROLLBACK.
+                    if let Err(first_rollback_err) = Self::rollback_transaction(&conn).await {
+                        // If we fail while rolling back, prioritize returning
```

Comment:
> What state are we in at this point? It seems like we failed a `ROLLBACK TO SAVEPOINT` and at that point. [The CockroachDB docs don't say anything about errors or failures here and I'm not sure if we're still "in" the nested transaction or what.](https://www.cockroachlabs.com/docs/v23.1/rollback-transaction)

It depends on how the `rollback_transaction` method has failed.

- Any time `batch_execute` is called -- which happens for **any** of these SQL operations (BEGIN, SAVEPOINT, ROLLBACK, COMMIT, etc) -- Diesel calls `update_transaction_manager_status` using the returned error code, which can possibly update the transaction state: https://docs.diesel.rs/master/src/diesel/pg/connection/mod.rs.html#135 
- The implementation of this function (https://docs.diesel.rs/master/src/diesel/pg/connection/mod.rs.html#232) basically matches on the observed transaction state from libpq and the error and can update Diesel's tracking in a handful of ways (including: doing nothing, indicating that the transaction manager is "broken", indicating that an error happened and we're expecting to roll all the way back to the top-level transaction).
- This is one of the ways we can call the function `set_requires_rollback_maybe_up_to_top_level` (https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#147-151) which allows SAVEPOINTS to be released without throwing subsequent errors, until we've finished rolling back to the top-level.

So, to answer your original question, "what state are we in", the answer is "it depends".

- The transaction depth *after* seeing this error will always be `1` from Diesel's perspective.
- We will always **tried** to have issued `ROLLBACK` after getting into this error case.

In some situations: the transaction manager will be "broken" and we'll expect all further Diesel operations to return: https://docs.diesel.rs/master/diesel/result/enum.Error.html#variant.BrokenTransactionManager

In other situations: the `ROLLBACK` error could have come from cockroachdb directly.

> Might we wind up returning from this function while still in the outermost transaction? edit: having look at the other thread, I see that `AnsiTransactionManager` tries to deal with stuff like this: https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#400-401

The Diesel interface tries really hard to ensure that the "transaction depth" is consistent in the success/failure cases -- basically, regardless of whether "rollback" or "commit" succeeds or fails, there are two outcomes:

1. The transaction "depth" has decreased by one.
2. The connection is permanently labelled as "broken" for unrecoverable errors (if you look in the Diesel source code, this is like a call to `set_in_error`). Once this happens, basically all subsequent diesel operations return a "broken transaction manager" error.

> Is that behavior a problem for us? e.g., is it possible that we try to do the `rollback to savepoint`, but diesel winds up doing that, it fails, and then it also rolls back the `BEGIN` and we've lost track of what depth we're at?

I don't think so - I think this would violate the aforementioned principle of "only change the transaction depth by at most one level".

> This is more of a general comment and I'm not sure what we can do about this: it feels like there's some critical state here (like "transaction depth" and "are we moving forward or backward") along with some guarantees made by various Diesel functions and async-bb8-functions, plus invariants maintained by the loop, etc. But I don't really understand what they all are. I don't think this is an async-bb8-diesel problem per se. But the net result is that as a reader I find it pretty hard to understand what's true at each point and have confidence in the code's correctness.
> 
> It might help [me?] to draw a state diagram. With that in hand, I wonder if an implementation could use this (maybe with something like [typestates](http://cliffle.com/blog/rust-typestate/)) to be more clearly, structurally correct. This sounds like a lot of work though and if we're all satisfied that this is correct as is it may not be worth it.

I agree that typestates could help here, but I'm not sure how much value they have in `async-bb8-diesel` alone -- it kinda sounds like the property you want extends into Diesel's implementation of transactions too? I have a fear of managing the "transaction depth" inside `async-bb8-diesel` via typestates, while a client using the API could just skip the `async-bb8-diesel` level and send requests directly through Diesel -- which would violate our higher-level expectations.

I started to write this up in https://docs.google.com/drawings/d/1JDU5-LsohXP7kEj_9ZFJUz64XyUXNJhZTu4J02sUYV4/edit?usp=sharing  -- just trying to explain what Diesel is doing -- but go figure, it's complicated. It wasn't written as a state machine, so this is basically an expansion of code logic into state machine form -- if we want something that looks like a cleaner state machine, we'll need to re-write Diesel's TransactionManager.

Reply:
I took a swing at making a state machine diagram for the CockroachDB client-side retry behavior.  I intended to finish it and compare it with the code, but haven't been able to prioritize it.  I figured I'd drop it here for posterity in case it's useful in the future.  (View source to see the underlyaing mermaid source.)

```mermaid
flowchart TD
    %% node definitions
    idle["connection idle"]
    begin["Execute `BEGIN`"]
    savepoint_start["Execute `SAVEPOINT`"]
    user_stuff["(consumer executes SQL queries)"]
    savepoint_release["Execute `RELEASE SAVEPOINT`"]
    commit["Execute `COMMIT`"]
    retry_decision{"Should retry?"}
    rollback_savepoint["Issue `ROLLBACK TO SAVEPOINT`"]
    rollback_all["Issue `ROLLBACK` (whole txn)"]
    dead["connection dead"]

    %% happy path
    idle -->|consumer calls `transaction_async_with_retry`| begin
    begin -->|ok| savepoint_start
    savepoint_start -->|ok| user_stuff
    user_stuff -->|ok| savepoint_release
    savepoint_release -->|ok| commit
    commit -->|ok| idle

    %% retry path
    user_stuff --> |fail| retry_decision
    retry_decision --> |yes| rollback_savepoint
    rollback_savepoint --> |ok| user_stuff
    savepoint_release --> |fail| retry_decision

    %% failure path
    retry_decision --> |no| rollback_all
    rollback_all --> |ok| idle
    commit --> |fail| rollback_all
    rollback_savepoint --> |fail| rollback_all

    %% other failures
    begin --> |fail| idle
    savepoint_start -->|fail| rollback_all

    %% really bad path
    rollback_all --> |fail| dead
```